File: python/problems/leetcode/0015_three_sum/solution.py
"""LeetCode 15: 3Sum."""

from __future__ import annotations


class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()
        n = len(nums)
        res: list[list[int]] = []
        for i in range(n):
            if i and nums[i] == nums[i - 1]:
                continue
            if nums[i] > 0:
                break
            l, r = i + 1, n - 1
            while l < r:
                s = nums[i] + nums[l] + nums[r]
                if s < 0:
                    l += 1
                elif s > 0:
                    r -= 1
                else:
                    res.append([nums[i], nums[l], nums[r]])
                    x, y = nums[l], nums[r]
                    while l < r and nums[l] == x:
                        l += 1
                    while l < r and nums[r] == y:
                        r -= 1
        return res


if __name__ == "__main__":
    out = Solution().threeSum([-1, 0, 1, 2, -1, -4])
    out_set = {tuple(t) for t in out}
    assert out_set == {(-1, -1, 2), (-1, 0, 1)}
    print("0015 OK")

File: cpp/problems/leetcode/0015_three_sum/solution.cpp
// LeetCode 15
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        int n = (int)nums.size();
        vector<vector<int>> res;
        for (int i = 0; i < n; ++i) {
            if (i && nums[i] == nums[i - 1]) continue;
            if (nums[i] > 0) break;
            int l = i + 1, r = n - 1;
            while (l < r) {
                long long s = (long long)nums[i] + nums[l] + nums[r];
                if (s < 0)
                    ++l;
                else if (s > 0)
                    --r;
                else {
                    res.push_back({nums[i], nums[l], nums[r]});
                    int xl = nums[l], yr = nums[r];
                    while (l < r && nums[l] == xl) ++l;
                    while (l < r && nums[r] == yr) --r;
                }
            }
        }
        return res;
    }
};

int main() {
    vector<int> a{-1, 0, 1, 2, -1, -4};
    auto out = Solution().threeSum(a);
    set<array<int, 3>> st;
    for (auto& v : out) st.insert({v[0], v[1], v[2]});
    assert(st.count({-1, -1, 2}) && st.count({-1, 0, 1}) && st.size() == 2u);
    cout << "0015 OK" << endl;
    return 0;
}
